fetch.js

```diff
@@ -0,0 +1,70 @@
+var Bluebird  = require('bluebird');
+var _defaults = require('lodash/object/defaults');
+
+var EmberCliDeployError = require('./errors/ember-cli-deploy-error');
+
+var redis = require('then-redis');
+var redisClient;
+var defaultConnectionInfo = {
+  host: "127.0.0.1",
+  port: 6379
+};
+
+var opts;
+var _defaultOpts = {
+  revisionQueryParam: 'index_key'
+};
+var _getOpts = function (opts) {
+  opts = opts || {};
+  return _defaults({}, opts, _defaultOpts);
+};
+
+var initialized = false;
+var _initialize = function (connectionInfo, passedOpts) {
+  opts = _getOpts(passedOpts);
+  var config = connectionInfo ? connectionInfo : defaultConnectionInfo;
+  redisClient = redis.createClient(config);
+
+  initialized = true;
+};
+
+var fetchIndex = function (req, appName, connectionInfo, passedOpts) {
+  if (!initialized) {
+    _initialize(connectionInfo, passedOpts);
+  }
+
+  var indexkey;
+  if (req.query[opts.revisionQueryParam]) {
+    var queryKey = req.query[opts.revisionQueryParam].replace(/[^A-Za-z0-9]/g, '');
+    indexkey = appName + ':' + queryKey;
+  }
+
+  var customIndexKeyWasSpecified = !!indexkey;
+  function retrieveIndexKey(){
+    if (indexkey) {
+      return Bluebird.resolve(indexkey);
+    } else {
+      return redisClient.get(appName + ":current").then(function(result){
+        if (!result) { throw new Error(); }
+        return result;
+      }).catch(function(){
+        throw new EmberCliDeployError("There's no " + appName + ":current revision. The site is down.", true);
+      });
+    }
+  }
+
+  return retrieveIndexKey().then(function(indexkey){
+    return redisClient.get(indexkey);
+  }).then(function(indexHtml){
+      if (!indexHtml) { throw new Error(); }
+      return indexHtml;
+  }).catch(function(err){
+    if (err.name === 'EmberCliDeployError') {
+      throw err;
+    } else {
+      throw new EmberCliDeployError("There's no " + indexkey + " revision. The site is down.", !customIndexKeyWasSpecified);
+    }
+  });
+};
+
+module.exports = fetchIndex;
```

Comment:
Meta comment on this file: This is where the naming convention for the redis keys are encoded. This implementation matches the original version of ember-deploy-redis and my talk. However, our thinking has evolved somewhat, and zapnito/ember-cli-deploy-redis uses a pattern like this: `appname:pagename:version` where "pagename" is generally `index`. Instead reading the current version from the `current` key and then reading that version, we're storing the contents of the current version in a `default` key, so we just read from that in the common case that no query param is sent.